sstables/sstables.cc

```diff
@@ -1603,6 +1619,9 @@ future<> sstable::load(sstables::foreign_sstable_open_info info) noexcept {
         validate_partitioner();
         return update_info_for_opened_data();
     }).then([this] {
+        return bytes_on_disk_uncached();
+    }).then([this] (uint64_t bytes) {
```

Comment:
I don't think we need this. At this point, bytes_on_disk() can return correct result since _metadata_size_on_disk + _data_file_size + _index_file_size are all set.

Reply:
"I can't use the existent bytes_on_disk() function for this purpose.
Namely, since at least commit https://github.com/scylladb/scylladb/commit/2dbae856f84ebef8267e53ee542cc9aa1da27775 ("sstable: Piggyback on sstable
parser and writer to provide bytes_on_disk", 2023-04-27), bytes_on_disk()
insists on "_data_file_size" and "_index_file_size" having been set (to
nonzero values). However, at least "_data_file_size" may not be set when
we call bytes_on_disk() from seal_sstable(), for example."

I think we can do it in open_data(). see that _manager.increment_total_reclaimable_memory_and_maybe_reclaim() happens there, once the sstable becomes ready for reads.

Reply:
@raphaelsc please see my comment [here](https://github.com/scylladb/scylladb/pull/20574#issuecomment-2347099183). `bytes_on_disk` is indeed fine for both `load` variants; that was what I tried first. `open_data` is listed there, as a participant in one of the `load` call trees; `open_data` populates `_data_file_size` and `_index_file_size` for that `load` variant. However:
1. You are [suggesting](https://github.com/scylladb/scylladb/pull/20574#discussion_r1761809170) not to modify the `load` variant that calls `open_data`. This may be a valid suggestion for that particular `load` variant, but then the recommendation about `open_data` seems confusing (it's a contradiction), because `open_data` is called from exactly that `load` variant, so the accounting would still happen there.
2. Furthermore, `open_data` is not called from any other location (at least inside the class). It is a public function though, and seems to be called from outside the class. All in all I can't determine what `open_data`'s official purpose is, and what general impact it would have if we added to accounting to `open_data`. What we need is a central (unique) entry point that stands for "discovery", but I'm not sure which sst member function that is.
3. Another objection (or question) is that `open_data` is never called on the write path. So we wouldn't account for disk usage right after a new sst is written, until it gets loaded (which may not happen for an indefinite time, IIUC). From your [original comment](https://github.com/scylladb/scylladb/issues/18191#issuecomment-2049050978), it seemed that `seal_sstable` was a good hook point on the write path. However, when I [tried to call](https://github.com/scylladb/scylladb/pull/20574#issuecomment-2347099183) `bytes_on_disk` from `seal_sstable`, it crashed, due to `_data_file_size` not having been set. So where do I set the index and data file sizes on the write path? (Again, `open_data` is not a part of that path.)
4. The above bullets are all about *incrementing* the stat (on the read path, and on the write path). However, we also need to decrement the stat; `unlink` seems like the place to do it. However, `unlink` can happen on the write path too (`ctor -> open_sstable -> seal_sstable -> close_files -> unlink`). And just as in `seal_sstable`, `_data_file_size` is unavailable in `unlink` on the write path, so I can't call `bytes_on_disk` from `unlink` on the write path.

Reply:
> I don't think we need this. At this point, bytes_on_disk() can return correct result since _metadata_size_on_disk + _data_file_size + _index_file_size are all set.
> I think we can do it in open_data().

Doesn't work; if I  modify both load variants to use bytes_on_disk(), boost/sstable_test fails (big_summary_query_0, big_summary_query_32, all_in_place).